bootloader/firmware/crc32.c

```diff
@@ -0,0 +1,69 @@
+#include "bootloader/firmware/crc32.h"
+#include "platform_specific_hal_conf.h"
+#include "common/firmware/errors.h"
+
+/**
+ * Used this to match results from python's zlib.crc32.
+ * https://m0agx.eu/2021/04/09/matching-stm32-hardware-crc-with-standard-crc-32/
+ */
+
+/**
+ * Handle to CRC module.
+ */
+static CRC_HandleTypeDef hcrc;
+
+/**
+ * Initialize CRC handle.
+ */
+static void MX_CRC_Init(void);
+
+
+/**
+ * Initialize crc module.
+ */
+void crc32_init() {
+    MX_CRC_Init();
+}
+
+/**
+ * Compute the CRC
+ * @param data Data
+ * @param length Length of data
+ * @return Computed CRC
+ */
+uint32_t crc32_compute(const uint8_t* data, uint8_t length) {
+    return ~HAL_CRC_Calculate(&hcrc, (uint32_t*)data, length);
+}
+
+/**
+ * Continue accumulating CRC using provided data.
+ * @param data Data
+ * @param length Length of data
+ * @return Accumulated CRC
+ */
+uint32_t crc32_accumulate(const uint8_t* data, uint8_t length) {
+    return ~HAL_CRC_Accumulate(&hcrc, (uint32_t*)data, length);
+}
+
+/**
+ * Reset the accumulated CRC value.
+ */
+void crc32_reset_accumulator() {
+    __HAL_CRC_DR_RESET(&hcrc);
+}
+
+/**
+ * Initialize the CRC unit.
+ */
+void MX_CRC_Init(void) {
+    hcrc.Instance = CRC;
+    hcrc.Init.DefaultPolynomialUse = DEFAULT_POLYNOMIAL_ENABLE;
```

Comment:
How did we decide on these CRC values?

Reply:
I did what I could to have the fw match what Python produces.

Reply:
Definitely looks good, the key to me is using a standard polynomial - everything else is just kind of matching bit ordering.